src/meta/src/stream/stream_graph/schedule.rs

```diff
@@ -0,0 +1,243 @@
+
+/// The internal distribution structure for processing in the scheduler.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+enum DistId {
+    Hash(HashMappingId),
+    Singleton,
+}
+
+/// Facts as the input of the scheduler.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+enum Fact {
+    /// An edge in the stream graph.
+    Edge {
+        from: Id,
+        to: Id,
+        dt: DispatcherType,
+    },
+    /// A distribution requirement for a building fragment.
+    InternalReq { id: Id, dist: DistId },
+    /// A distribution requirement for an existing fragment.
+    ExternalReq { id: Id, dist: DistId },
+    /// The default distribution for a building fragment.
+    Default(DistId),
+}
+
+crepe::crepe! {
+    @input
+    struct Input(Fact);
+
+    struct Edge(Id, Id, DispatcherType);
+    struct Internal(Id, DistId);
+    struct External(Id, DistId);
+    struct Default(DistId);
+    struct Fragment(Id);
+
+    struct Requirement(Id, DistId);
+
+    @output
+    struct Success(Id, DistId);
+    @output
+    #[derive(Debug)]
+    struct Failed(Id);
+
+    // Extract facts.
+    Edge(from, to, dt) <- Input(f), let Fact::Edge { from, to, dt } = f;
+    Internal(id, dist) <- Input(f), let Fact::InternalReq { id, dist } = f;
+    External(id, dist) <- Input(f), let Fact::ExternalReq { id, dist } = f;
+    Default(dist) <- Input(f), let Fact::Default(dist) = f;
+
+    // Internal fragments.
+    Fragment(x) <- Edge(x, _, _), !External(x, _);
+    Fragment(y) <- Edge(_, y, _), !External(y, _);
+
+    // Requirements in the facts.
+    Requirement(x, d) <- Internal(x, d);
+    Requirement(x, d) <- External(x, d);
+    // Requirements of `NoShuffle` edges.
+    Requirement(x, d) <- Edge(x, y, NoShuffle), Requirement(y, d);
+    Requirement(y, d) <- Edge(x, y, NoShuffle), Requirement(x, d);
+    // Requirements of `Simple` edges.
+    Requirement(y, DistId::Singleton) <- Edge(_, y, Simple);
+
+    // Multiple requirements lead to failure.
+    Failed(x) <- Requirement(x, d1), Requirement(x, d2), (d1 != d2);
+
+    // Take the single requirement as the result.
+    Success(x, d) <- Fragment(x), Requirement(x, d), !Failed(x);
+    // Take the default distribution as the result, if no requirement.
+    Success(x, d) <- Fragment(x), Default(d), !Requirement(x, _);
+}
+
+/// The distribution of a fragment.
+#[derive(EnumAsInner)]
+pub(super) enum Distribution {
+    Singleton,
+    Hash(ParallelUnitMapping),
+}
+
+impl Distribution {
+    /// The parallelism required by the distribution.
+    pub fn parallelism(&self) -> usize {
+        match self {
+            Distribution::Singleton => 1,
+            Distribution::Hash(mapping) => mapping.iter_unique().count(),
+        }
+    }
+}
+
+/// [`Scheduler`] schedules the distribution of fragments in a stream graph.
+pub(super) struct Scheduler {
+    /// The default distribution for fragments, if there's no requirement derived.
+    default_hash_mapping: ParallelUnitMapping,
+}
```

Comment:
There are two `Scheduler`s for streaming now. One is used to schedule the distribution of fragments in a stream graph. The other one is used to set up location and `vnode_bitmap` for actors. Will we unify them in the future? Do they need new names to be distinguished from one another?

Reply:
The original one will be removed, as the parallel unit mapping of the fragment here actually decides the locations of the expanded actors. The `Actor` built from the `StreamGraphBuilder` will directly be complete and there'll be no extra step of filling the `vnode_bitmap` for actors.

Reply:
Cool, lots of scheduler complexity added by delta join could be solved elegantly by your refactoring.

Reply:
It seems like this method could be applied to scale as well.🤔

Reply:
Good catch. 🥺 I'm also thinking of whether this can be used to simplify the resolving of `NoShuffle` for scaling and make it more general.